### PR TITLE
tools/rados: support regular expression in rados ls command to filter

### PR DESCRIPTION
**Pull request by liupan1111**
object name.

There will be millions of objects in the backend, and rados ls will be very time consuming, and waste throughput when hight network pressure. The customers ask to provide a method to filter objects by name. 

image
When use regular expression:
image

Signed-off-by: Pan Liu <wanjun.lp@alibaba-inc.com>

### CODE REVIEW

**Comment by liupan1111** [comment]
@dillaman could help take a look? Thanks!

**Comment by dillaman** [comment]
@liupan1111 I'm confused -- why not follow the standard *nix convention of chaining commands together a la ```rados -p rbd ls | grep '.*image"```?

**Comment by liupan1111** [comment]
@dillaman The reason is:
If the user tries: rados -p rbd ls | grep '.*image", it means all objects(millions) in rbd pool should return to client first, and grep it. As I described in the description of this PR:

> and rados ls will be very time consuming, and waste throughput when hight network pressure. The customers ask to provide a method to filter objects by name.

By using this regular expression feature, this job will be done in the backend(OSD side), will not consume the CPU of client, and only pass the matched objects to the user.

**Comment by liupan1111** [comment]
@dillaman If the name length of objects on average is 10 byte, and 1m objects, the total throughput will be 10Gb. But maybe only a small part of these objects are needed by the user.

**Comment by dillaman** [comment]
@liupan1111 ACK -- I didn't look at it before commenting. You should break it up into two commits (for the OSD and tool changes) and make sure you properly tag this PR w/ ```core``` as well

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -769,6 +790,20 @@ class PGLSParentFilter : public PGLSFilter {
                       bufferlist& outdata) override;
 };
 
+bool PGLSNameFilter::filter(const hobject_t &obj,
+                              bufferlist& indata, bufferlist& outdata)
+{
+  // match user regex
+  if (regex != "" && !regex.empty() && regex != "no_reg") {
```
Comment (dillaman) [inline src/osd/PrimaryLogPG.cc]: Don't use "magic" values -- if the regex isn't going to be used, don't use the filter.

Comment (branch-predictor) [inline src/osd/PrimaryLogPG.cc]: Put that in init() function, so it doens't need to be executed for each object. Doing this on a pool with 1M objects will certainly have performance impact on that OSD.

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -769,6 +790,20 @@ class PGLSParentFilter : public PGLSFilter {
...
+    std::string obj_name = obj.oid.name;
+    std::regex e(regex);
```
Comment (branch-predictor) [inline src/osd/PrimaryLogPG.cc]: That also should go into init() function. Construction of regex object includes parsig and compiling an regex, and these aren't free.

**Comment by liupan1111** [comment]
> @liupan1111 ACK -- I didn't look at it before commenting. You should break it up into two commits (for the OSD and tool changes) and make sure you properly tag this PR w/ core as well

Yup!

File: src/osd/PrimaryLogPG.cc (hunk 1 above)
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: ok

Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: @branch-predictor I guess this check could be removed if I follow the comment from Jason.

> Don't use "magic" values -- if the regex isn't going to be used, don't use the filter.

File: src/osd/PrimaryLogPG.cc (hunk 2 above)
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: yes, thanks!

**Comment by liupan1111** [comment]
made mistake when rebase... so please ignore new commit. I will modify as your suggestions ASAP.

**Comment by liupan1111** [comment]
@dillaman @branch-predictor updated as your suggestion, please help take a look.

File: src/osd/PrimaryLogPG.cc [hunk 3]
```diff
@@ -727,6 +729,31 @@ void PrimaryLogPG::maybe_force_recovery()
...
 
+class PGLSNameFilter : public PGLSFilter {
+  std::regex reg;
+public:
+  int init(bufferlist::iterator &params) override
+  {
+    try {
+      string regex;
+      ::decode(regex, params);
+      if (regex == "" || regex.empty()) {
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: what's the difference between `regex == ""` and `regex.empty()`?

File: src/osd/PrimaryLogPG.cc [hunk 4]
```diff
@@ -727,6 +729,31 @@ void PrimaryLogPG::maybe_force_recovery()
...
+        throw buffer::error();
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: no need to throw an exception here. please just return `-EINVAL` instead.

File: src/osd/PrimaryLogPG.cc [hunk 5]
```diff
@@ -727,6 +729,31 @@ void PrimaryLogPG::maybe_force_recovery()
     maybe_kick_recovery(soid);
 }
 
+class PGLSNameFilter : public PGLSFilter {
+  std::regex reg;
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: `reg` deserves a better name. use `regex` if you please.

File: src/osd/PrimaryLogPG.cc [hunk 6]
```diff
@@ -727,6 +729,31 @@ void PrimaryLogPG::maybe_force_recovery()
...
+        reg = regex;
+      }
+    } catch (buffer::error &e) {
+      return -EINVAL;
+    }
+
+    return 0;
+  }
+
+  virtual ~PGLSNameFilter() {}
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: no point to have an empty dtor explicitly.

File: src/osd/PrimaryLogPG.cc [hunk 7]
```diff
@@ -769,6 +796,17 @@ class PGLSParentFilter : public PGLSFilter {
                       bufferlist& outdata) override;
 };
 
+bool PGLSNameFilter::filter(const hobject_t &obj,
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: would be better to move the implementation of this method up to where it is declared instead of interleaving it with another class.

File: src/osd/PrimaryLogPG.cc [hunk 8]
```diff
@@ -769,6 +796,17 @@ class PGLSParentFilter : public PGLSFilter {
...
+                              bufferlist& indata, bufferlist& outdata)
+{
+  std::string obj_name = obj.oid.name;
+  // match user regex
+  if (!std::regex_match (obj_name, reg)) {
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: please remove the comment, and put:
```c++
return std::regex_match(obj.oid.name, reg);
```
directly.

File: src/osd/PrimaryLogPG.cc [hunk 9]
```diff
@@ -54,6 +54,8 @@
 #include "include/assert.h"  // json_spirit clobbers it
 #include "include/rados/rados_types.hpp"
 
+#include <regex>
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: please use `boost::regex`. see http://tracker.ceph.com/issues/20112

Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: please use `osd: ` as the prefix of the commit message's title. no need to include "src".

File: src/tools/rados/rados.cc [hunk 10]
```diff
@@ -1656,6 +1659,10 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
   if (i != opts.end()) {
     pool_name = i->second.c_str();
   }
+  i = opts.find("regex");
+  if (i != opts.end()) {
+    regex = i->second.c_str();
```
Comment (tchaikov) [inline src/tools/rados/rados.cc]: could check if it's empty here.

File: src/tools/rados/rados.cc [hunk 11]
```diff
@@ -2135,8 +2142,16 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
...
+        librados::NObjectIterator i;
+	if (regex == NULL) {
+          i = io_ctx.nobjects_begin();
+        } else {
+          bufferlist filter_bl;
+          ::encode("name", filter_bl);
+          ::encode(regex, filter_bl);
+          i = io_ctx.nobjects_begin(filter_bl);
+        }
+        librados::NObjectIterator i_end = io_ctx.nobjects_end();
```
Comment (tchaikov) [inline src/tools/rados/rados.cc]: use `auto` if you please.

File: doc/man/8/rados.rst [hunk 12]
```diff
@@ -106,8 +106,8 @@ Pool specific commands
 :command:`listwatchers` *name*
   List the watchers of object name.
 
-:command:`ls` *outfile*
-  List objects in given pool and write to outfile.
+:command:`ls` *outfile* --regex *regular expression*
```
Comment (tchaikov) [inline doc/man/8/rados.rst]: please add square brackets to note that `--regex` option is optional. please don't add space in the metavar, `--regex *pattern*` is good enough

File: doc/man/8/rados.rst [hunk 13]
```diff
@@ -106,8 +106,8 @@ Pool specific commands
...
+  List objects in given pool and write to outfile. Regular expression could be passed in to filter object names.
```
Comment (tchaikov) [inline doc/man/8/rados.rst]: Only the matched objects are listed if *--regex* option is specified.

File: qa/workunits/rados/test_rados_tool.sh [hunk 14]
```diff
@@ -381,6 +381,11 @@ test_ls() {
     then
         die "Created $OBJS objects in default namespace but saw $CHECK"
     fi
+    CHECK=$("$RADOS_TOOL" -p $p ls --regex obj.* 2> /dev/null | wc -l)
```
Comment (tchaikov) [inline qa/workunits/rados/test_rados_tool.sh]: please quote `obj.*` like `"obj.*"`, to prevent the shell from expanding it. also, better off specifying a pattern other than `obj.*` like `obj1.*` and update the `if` clause accordingly, so this feature can be tested in a more discriminative way.

File: src/osd/PrimaryLogPG.cc (hunk 9 above)
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: cool, thanks for remind.

Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: ok

File: src/osd/PrimaryLogPG.cc (hunk 5 above)
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: ok

File: src/osd/PrimaryLogPG.cc (hunk 3 above)
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: regex.empty is enough in my mind.

File: src/osd/PrimaryLogPG.cc (hunk 4 above)
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: ok

File: src/tools/rados/rados.cc (hunk 11 above)
Comment (liupan1111) [inline src/tools/rados/rados.cc]: er... do you mean "auto i_end = ...." ?

File: doc/man/8/rados.rst (hunk 13 above)
Comment (liupan1111) [inline doc/man/8/rados.rst]: ok

File: doc/man/8/rados.rst (hunk 12 above)
Comment (liupan1111) [inline doc/man/8/rados.rst]: it is true.

File: qa/workunits/rados/test_rados_tool.sh (hunk 14 above)
Comment (liupan1111) [inline qa/workunits/rados/test_rados_tool.sh]: good point.

File: src/osd/PrimaryLogPG.cc (hunk 7 above)
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: ok

File: src/osd/PrimaryLogPG.cc (hunk 8 above)
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: ok

File: src/tools/rados/rados.cc (hunk 11 above)
Comment (liupan1111) [inline src/tools/rados/rados.cc]: @tchaikov

**Comment by liupan1111** [comment]
@tchaikov updated as your comment.

File: src/osd/PrimaryLogPG.cc [hunk 15]
```diff
@@ -727,6 +728,37 @@ void PrimaryLogPG::maybe_force_recovery()
...
+      } else {
+        regex = reg_exp;
+      }
+    } catch (buffer::error &e) {
+      return -EINVAL;
+    }
+
+    return 0;
+  }
+  ~PGLSNameFilter() override {}
```
Comment (dillaman) [inline src/osd/PrimaryLogPG.cc]: Nit: not needed

File: src/tools/rados/rados.cc [hunk 16]
```diff
@@ -159,6 +159,8 @@ void usage(ostream& out)
 "GLOBAL OPTIONS:\n"
 "   --object_locator object_locator\n"
 "        set object_locator for operation\n"
+"   --regex regular expression\n"
```
Comment (dillaman) [inline src/tools/rados/rados.cc]: Nit: this isn't really a global option -- it's only valid for ```ls```

File: src/osd/PrimaryLogPG.cc [hunk 17]
```diff
@@ -727,6 +728,37 @@ void PrimaryLogPG::maybe_force_recovery()
...
+public:
+  int init(bufferlist::iterator &params) override
+  {
+    try {
+      string reg_exp ;
+      ::decode(reg_exp, params);
+      if (reg_exp.empty()) {
+        return -EINVAL;
+      } else {
+        regex = reg_exp;
```
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: agree

File: src/osd/PrimaryLogPG.cc (hunk 15 above)
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: The other two Filters have same destructors, do you want me to delete them all?

Comment (dillaman) [inline src/osd/PrimaryLogPG.cc]: No need to change existing code -- but definitely no need to add one to new code where it isn't needed.

File: src/tools/rados/rados.cc (hunk 16 above)
Comment (liupan1111) [inline src/tools/rados/rados.cc]: yes.

File: src/osd/PrimaryLogPG.cc (hunk 15 above)
Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: yup

File: src/tools/rados/rados.cc (hunk 11 above)
Comment (tchaikov) [inline src/tools/rados/rados.cc]: yes.

File: src/osd/PrimaryLogPG.cc [hunk 18]
```diff
@@ -746,6 +778,18 @@ class PGLSPlainFilter : public PGLSFilter {
                       bufferlist& outdata) override;
 };
 
+bool PGLSPlainFilter::filter(const hobject_t &obj,
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: please leave the existing code as it is, unless you are touching it also.

Comment (liupan1111) [inline src/osd/PrimaryLogPG.cc]: ok...

File: src/osd/PrimaryLogPG.cc [hunk 19]
```diff
@@ -727,6 +728,37 @@ void PrimaryLogPG::maybe_force_recovery()
...
+      ::decode(reg_exp, params);
+      if (reg_exp.empty()) {
+        return -EINVAL;
+      } else {
+        regex = reg_exp;
+      }
+    } catch (buffer::error &e) {
+      return -EINVAL;
+    }
+
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc]: nit, remove the empty line.

**Comment by liupan1111** [comment]
@dillaman @tchaikov updated.

File: src/osd/PrimaryLogPG.cc [hunk 20]
```diff
@@ -727,6 +728,35 @@ void PrimaryLogPG::maybe_force_recovery()
...
+      } else {
+        regex.assign(reg_exp, boost::regex_constants::no_except);
+      }
+    } catch (buffer::error &e) {
+      return -EINVAL;
+    }
+    return 0;
+  }
+  bool filter(const hobject_t &obj, bufferlist& indata,
+                            bufferlist& outdata) override;
```
Comment (dillaman) [inline src/osd/PrimaryLogPG.cc]: Nit: indentation spacing ... and since the function is really only one line long, why not just inline the definition here it while you're at it?

File: qa/workunits/rados/test_rados_tool.sh [hunk 21]
```diff
@@ -381,6 +381,11 @@ test_ls() {
     then
         die "Created $OBJS objects in default namespace but saw $CHECK"
     fi
+    CHECK=$("$RADOS_TOOL" -p $p ls --regex "obj1.*" 2> /dev/null | wc -l)
+    if [ "$CHECK" -ne 11 ];
+    then
+        die "Created 2 objects with the name matached 'obj1.*' in default namespace but saw $CHECK"
```
Comment (dillaman) [inline qa/workunits/rados/test_rados_tool.sh]: Nit: 2 --> 11?

Comment (liupan1111) [inline qa/workunits/rados/test_rados_tool.sh]: yes, 11

**Comment by liupan1111** [comment]
@dillaman updated.

**Comment by liupan1111** [comment]
There is similar as ls huge number of files in a filesystem dir, yes, it is not efficient, but linux should also provide such kind of function.

The use case for our user is: use app to call librados-java(which will call librados) to write nearly more than 1m objects, and regular expression of object names is asked to reduce the compact of network throughput.

There is similar as ls huge number of files in a filesystem dir, yes, it is not efficient, but linux should also provide such kind of function, which includes regular expression.

If you are worry about the user may mislead "anybody think object listing is efficient", we could remind this in doc/commit description/PR description.

**Comment by liewegas** [comment]
Why is the application listing objects, though?  The concern is that applications usually shouldn't be designed around using the pgls interface if they can avoid it because it is still pretty inefficient (even with a regex match on the server side).

I think noting that in the docs for the regex feature, and also in the rados CLI help (that listing in inefficient even with a regex) would help...

**Comment by branch-predictor** [comment]
I agree with @gregsfortytwo. From my experience, if there's some functionality that can be abused, there will be someone who does just that and that's what is happening here.
@liupan1111 If that's your application that writes these objects, why not cache the names of objects after they're written instead of relying on Ceph? You don't need a database for that, you can cache it in memory and re-populate it with "ls" later (for example, during startup). It'll cost you memory, sure, but it'll be less costly than reiterating over the object list again and again.

**Comment by gregsfortytwo** [comment]
Listing RADOS objects is fundamentally different than listing a large directory. You expect filesystems to provide that ability well; you do not expect it of most object stores (certainly not RADOS). Listing a directory involves streaming block reads off disk, and in some filesystems large dirs can be inefficient because of the sorting you apply (or if you demand full file detail and they have to run off and get the whole inode instead of just the directory), but in RADOS the worst case is not uncommon and is every object requiring a random IO.

Basically, the existing users of object listing interfaces are for disaster recovery operations where we actually *want* to enumerate every object in the system because we've lost a random selection of objects, so we don't have all our indexes and can't trust the fragments we have anyway. None of the examples here look like that, and I can't imagine a scenario where a regex is really what you want for disaster recovery anyway.
